test/unit/math/rev/mat/prob/neg_binomial_2_log_glm_lpmf_test.cpp

```diff
@@ -0,0 +1,164 @@
+#include <stan/math/rev/mat.hpp>
+#include <gtest/gtest.h>
+#include <stan/math/prim/scal/fun/value_of.hpp>
+
+
+using stan::math::var;
+using Eigen::Dynamic;
+using Eigen::Matrix;
+
+//  We check that the values of the new regression match those of one built
+//  from existing primitives.
+TEST(ProbDistributionsNegBinomial2LogGLM,
+     glm_matches_neg_binomial_2_log_doubles) {
+  Matrix<int, Dynamic, 1> n(3, 1);
+  n << 1, 0, 1;
+  Matrix<double, Dynamic, Dynamic> x(3, 2);
+  x << -12, 46, -42,
+      24, 25, 27;
+  Matrix<double, Dynamic, 1> beta(2, 1);
+  beta << 0.3, 2;
+  double alpha = 0.3;
+  Matrix<double, Dynamic, 1> alphavec = alpha * Matrix<double, 3, 1>::Ones();
+  Matrix<double, Dynamic, 1> theta(3, 1);
+  theta = x * beta + alphavec;
+  Matrix<double, Dynamic, 1> phi(3, 1);
+  phi << 2, 1, 0.2;
+
+  EXPECT_FLOAT_EQ((stan::math::neg_binomial_2_log_lpmf(n, theta, phi)),
```

Comment:
This pattern is looking so familiar! I wonder if there's a way to abstract further (not in this PR, just brainstorming). Like maybe using templates to construct a GLM version of any distribution... with C++11 we can have variadic template args and fun things like that. Curious if you have thoughts here.

Reply:
I agree. This could be done much more generally. I even wonder how much automation could be put into constructing compiler optimisations of the shape f(g(x)) -> f_composed_with_g(x). Ideally, you'd want to just scan a large base of Stan code, see which primitives are most often combined and then automatically implement a primitive for the composite function. That's a bit of a wild idea of course as it requires us to be able to compute derivatives symbolically. But I wonder if we could automate some of it in cases like GLMs like you suggest.

Reply:
Right - I mean there are symbolic differentiation libraries that are pretty good, though I'm not sure they're great 100% of the time. We might need some kind of heuristic for deciding during compilation whether 1) calculating the symbolic version is likely to bear fruit 2) given a symbolic version, if it's likely to be faster than autodiff.